### PR TITLE
feat: Implement EC2 Key Pairs (#1403)

### LINKED ISSUE #1403: Support EC2 Key Pairs
### Describe the resource.

https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.DescribeKeyPairs

DescribeKeyPairs API operation for Amazon Elastic Compute Cloud.

Describes the specified key pairs or all of your key pairs.

### Use Case

Audit EC2 key pairs for deny-listed fingerprints. 

### Additional context

It might be good to have the `aws_ec2_instances` table also include the fingerprint. We can join on the key_name though. It might also be good to pull the public key material if there's a use case for it.

### PR DESCRIPTION
**Pull request by jhudson10x**
#### Summary

Add new resource for EC2 Key Pairs

https://github.com/cloudquery/cq-provider-aws/issues/1403

---

Use the following steps to ensure your PR is ready to be reviewed

- [X] Read the [contribution guidelines](../blob/main/CONTRIBUTING.md) 🧑‍🎓
- [X] Run `go fmt` to format your code 🖊
- [X] Lint your changes via `golangci-lint run` 🚨 (install golangci-lint [here](https://golangci-lint.run/usage/install/#local-installation))
- [X] Update or add tests. Learn more about testing [here](https://docs.cloudquery.io/docs/developers/sdk/testing) 🧪
- [X] Update the docs by running `go run ./docs/docs.go` and committing the changes 📃
- [ ] Ensure the status checks below are successful ✅

### CODE REVIEW

File: resources/services/ec2/key_pairs.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+import (
+	"context"
+
+	"github.com/aws/aws-sdk-go-v2/service/ec2"
+	"github.com/cloudquery/cq-provider-aws/client"
+	"github.com/cloudquery/cq-provider-sdk/provider/diag"
+	"github.com/cloudquery/cq-provider-sdk/provider/schema"
+)
+
+func Ec2KeyPairs() *schema.Table {
```
Comment (candiduslynx) [inline resources/services/ec2/key_pairs.go:12]: Kindly add `cq-gen` config for this resource.
Additionally, the generate instruction comment should be present here, and no diff should occur once you run this generate command on this file.

See [AWS SES Template resource](https://github.com/cloudquery/cq-provider-aws/blob/main/resources/services/ses/templates.go#L13) for inspiraition.

File: resources/services/ec2/key_pairs.go [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+	return &schema.Table{
+		Name:          "aws_ec2_key_pairs",
+		Description:   "Describes an EC2 Key Pair.",
+		Resolver:      fetchEc2KeyPairs,
+		Multiplex:     client.ServiceAccountRegionMultiplexer("ec2"),
+		IgnoreError:   client.IgnoreWithInvalidAction,
```
Comment (candiduslynx) [inline resources/services/ec2/key_pairs.go:18]: Why did you choose this specific ignore function?
Will `client.client.IgnoreAccessDeniedServiceDisabled` work?

**Review by candiduslynx — CHANGES_REQUESTED** [review]

File: resources/services/ec2/key_pairs.go [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		DeleteFilter:  client.DeleteAccountRegionFilter,
+		Options:       schema.TableCreationOptions{PrimaryKeys: []string{"key_name"}},
```
Comment (candiduslynx) [inline resources/services/ec2/key_pairs.go:20]: Use `key_pair_id` instead

**Review by candiduslynx — CHANGES_REQUESTED** [review]

File: resources/services/ec2/key_pairs.go (hunk 2 above)
Comment (jhudson10x) [reply resources/services/ec2/key_pairs.go:18]: I don't have a good reason for having chosen that specific ignore function. I will change it to client.IgnoreAccessDeniedServiceDisabled. If there is an even better ignore handler to use for this situation tell me and I will change it again.

File: resources/services/ec2/key_pairs.go (hunk 1 above)
Comment (jhudson10x) [reply resources/services/ec2/key_pairs.go:12]: I will add an entry to the `resources/services/ec2/gen.hcl` file for this resource. I notice that there are other resources with missing entries in this file. Can you provide me more info about the purpose of this file?

Comment (candiduslynx) [reply resources/services/ec2/key_pairs.go:13]: We tend to use `cq-gen` to generate resources (may be installed via `make install-tools`).
Additionally, it provides an easy way to specify the resource structure.

Also, we are going to check for drift changes for the generated resources (so you should be able to run `go generate` on this file and see no changes whatsoever).

File: resources/services/ec2/gen.hcl [hunk 4]
```diff
@@ -249,4 +249,44 @@ resource "aws" "ec2" "instance_types" {
...
+  multiplex "AwsAccountRegion" {
+    path   = "github.com/cloudquery/cq-provider-aws/client.ServiceAccountRegionMultiplexer"
+    params = ["ec2"]
+  }
+
+  options {
+    primary_keys = ["key_pair_id"]
+  }
+
+  userDefinedColumn "account_id" {
```
Comment (candiduslynx) [inline resources/services/ec2/gen.hcl:271]: does Key Pair have ARN?

Comment (candiduslynx) [reply resources/services/ec2/gen.hcl:271]: Then you should add a resolver for it, too.

Basically, you should add (and specify this column to e primary key, see lots of examples in the reources):
```hcl
userDefinedColumn "arn" {
    type              = "string"
    description       = "The Amazon Resource Name (ARN) of the Key Pair"
    generate_resolver = true
}
```
Then you'll need to regenerate the implementation and implement the newly generated resolver.
You can find examples of resolvers utilizing helper functions like [this one](https://github.com/cloudquery/cq-provider-aws/blob/main/client/helpers.go#L290) throughout the codebase.

Comment (jhudson10x) [reply resources/services/ec2/gen.hcl:271]: The func signature that cq-gen has a return type of `error`. Should I change the sig to return `schema.ColumnResolver` like the example you linked, or should I keep the generated sig and do like `resources/services/iot/iot_topic_rules.go#L1337` and write into the resource?

Comment (candiduslynx) [reply resources/services/ec2/gen.hcl:271]: You should do something like https://github.com/cloudquery/cq-provider-aws/blob/main/resources/services/ec2/images.go#L42-L44

I suppose you can just fill in the proper func body for the generated func.

**Comment by erezrokah** [comment]
Hi @jhudson10x and @candiduslynx, I moved this PR to our monorepo (including original commits) via https://github.com/cloudquery/cloudquery/pull/1325. Since https://github.com/cloudquery/cloudquery/pull/1325 was approved and merged I suggest to do any follow up changes in another PR.
Thanks for the contribution @jhudson10x!

**Comment by erezrokah** [comment]
Closing in favor https://github.com/cloudquery/cloudquery/pull/1325

**Pull request closed without merging**